File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int bfs(int start, vector<vector<int>>& graph, vector<bool>& visited) {
    queue<int> q;
    q.push(start);
    visited[start] = true;
    int count = 1;
    
    while(!q.empty()) {
        int node = q.front();
        q.pop();
        for(int next : graph[node]) {
            if(!visited[next]) {
                visited[next] = true;
                q.push(next);
                count++;
            }
        }
    }
    return count;
}

int solution(int n, vector<vector<int>> wires) {
    int answer = n;
    
    for (int i = 0; i < wires.size(); i++) {
        vector<vector<int>> graph(n + 1);
        for(int j = 0; j < wires.size(); j++) {
            if(i == j) continue;
            int a = wires[j][0], b = wires[j][1];
            graph[a].push_back(b);
            graph[b].push_back(a);
        }
        
        vector<bool> visited(n + 1, false);
        int cnt = bfs(1, graph, visited);
        int diff = abs((n - cnt) - cnt);
        answer = min(answer, diff);
    }
    
    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        int s = wires.length;
        
        for (int i = 0; i < s; i++) {
            ArrayList<ArrayList<Integer>> edgeList = new ArrayList<>();
            
            for (int j = 0; j <= n; j++) {
                edgeList.add(new ArrayList<>());
            }
            
            for (int j = 0; j < s; j++) {
                if (j == i) continue; 
                edgeList.get(wires[j][0]).add(wires[j][1]);
                edgeList.get(wires[j][1]).add(wires[j][0]);
            }
            
            boolean[] visited = new boolean[n + 1];
            int cnt = dfs(1, visited, edgeList);
            
            answer = Math.min(answer, Math.abs(cnt - n + cnt));
        }
        
        return answer;
    }
    
    private int dfs(int start, boolean[] visited, ArrayList<ArrayList<Integer>> edgeList) {
        int cnt = 1;
        visited[start] = true;
        for (int nxt : edgeList.get(start)) {
            if (visited[nxt]) continue;
            cnt += dfs(nxt, visited, edgeList);
        }
        return cnt;
    }
}